### ISSUE TITLE
Allow access to a new `/organizations` endpoint

### ISSUE THREAD

**Issue by slifty** [PhilanthropyDataCommons/service#777]
This new endpoint would make it possible to access organization data as both a public user AND a logged in user (aka superuser at this point).

In order to implement it we'll need to add in a few concepts to the schema:

1. The concept of an organization (this is not a trivial concept to get right, and is essentially a replacement of the concept of an applicant see #730).

2. The concept of scope to base fields (organization vs proposal for now)

3. The concept of "public" base fields

Once this is all done, it'll be possible to return a list of organizations along with any "organization" base field (narrowed to include / exclude non-public data depending on the user's auth)

This issue also replaces #767

**Comment by slifty** [comment]
I can already tell this issue is gonna be bigger than I thought when opening it BUT *cracks knuckles* let's dive in...

We currently have the concept of an applicant, which we now know is both too specific and too broad.

Here's what I mean:

* Applicants cannot be individuals at this stage of the PDC.  We're only collecting organization data.
* Organizations might not be applicants!  For instance, maybe we have an organization in the PDC that came from a third party data provider.
* Our ERD had a baked-in assumption that there will be exactly one applicant for a given proposal.  In reality there might be more than one applicant.

There is also the question of "what data defines an organization" -- to put this in data modeling terms, what would a natural primary key for an organization be?  We know it is *not* just the EIN since that does not account for complex hierarchies (e.g. the Red Cross) which might have multiple sub-organizations within one parent organization which independently apply for grants.

Let's take this in pieces...

## Question 1: Applicant vs Organization

I think we already have consensus on this: drop applicant, create organization.  We would also remove `applicantId` from the proposal table and this in turn would allow us to update bulk upload to no longer require any particular data about applicants (e.g. applicant email address).

## Question 2: Organization "natural identifiers"

From our discussion so far I think there are three possible attributes to a natural identifier:

1. EIN (Employee Identification Number)
2. Organization Name (This *could* actually be split into two fields of "Organization name" and "branch name" where branch name is null most of the time)
3. Organization Address (which is more than one field but that's not important here...)

I think the cleanest natural key would be EIN + Org Name + Branch Name, however I also don't think the concept of "branch name" exists anywhere.  At the very least it would not consistently exist in application forms.

For that reason I'd like to lean in on EIN + Organization address.

IN THE SHORT TERM this would mean that if an organization changes their address, they would get two separate organizations within the PDC.  That's *not* ideal, but would be a natural consequence.

IN THE LONG TERM we are going to want / need a way to establish relationships between organizations within the PDC (e.g. parent / child relationships), and one of those relationships could be something like "alias" -- which would allow us to represent equivalence classes between organizations even if they have different addresses.

So in summary I'm proposing that EIN + Organization Address become the "identifying" features of an organization

## Question 3: Associating proposals with organizations

When a proposal is added how do we link it to organizations?  If there are multiple organizations associated with a proposal how do we link to THOSE organizations?  If there are multiple organizations associated with a proposal, which field values end up applying to those organizations and how do we resolve that association?

These are some life's biggest mysteries.

I'll have to come back to question 3.

**Comment by slifty** [comment]
Ok back to question 3!

I'm thinking the right long term approach is that proposals and organizations have a many-to-many relationship.  The join table would also support different types of relationship -- e.g. something like "primary applicant" and "secondary applicant".

The key business logic is that organizational field values would apply to primary applicants' values when the organization is looked up, but not secondary (e.g. if a proposal had "org name" of "Foo Inc." then the organization that is associated as a primary applicant would have "Foo Inc." as a field value for org name, whereas an organization associated as a secondary applicant would NOT)

For the initial implementation I don't think we need to support secondary applicants, and for that reason I'd like to propose that all joins are primary applicants.  Eventually this model can be expanded to support more complexity.

I *do* still think we should still use a join table, even though the previous simplifying assumption technically means our first implementation is 1-to-many (which would not normally require a join table).

------

TL;DR, the order of operations for this issue is

1. Remove the `applicants` table (and any primary keys that point to it).
2. Create a new `organizations` table.
3. Create a new `organizations_proposals` join table.
4. Implement the new `/organizations` endpoint.

The infomediary data we have loaded was loaded in an intentionally hacky way, and I think this is the point that we must pay the piper for that decision.  The infomediary data will no longer be linked to an applicantId but will still exist in the database.  It's not clear to me if this impacts the front end yet.

**Comment by slifty** [comment]
One issue with using `address` as part of the identifying key is that addresses are notoriously non-standardized.

For that reason I'm shifting my plan of using address, and instead it will be EIN + Organization Name.

It seems reasonable that a "child" organization sharing an EIN might populate with a different organization name than the parent (e.g. "Red Cross Internationa" vs "Red Cross Philadelphia").

**Comment by slifty** [comment]
There is more work to be done on the `/organizations` endpoint (adding base field data in particular) but I'm going to move those into separate issues.  The stated opening scope of this issue is complete!
